File: euler/P21.java
// Project Euler, problem 21: Amicable numbers (http://projecteuler.net/problem=21). 
// Let d(n) be defined as the sum of proper divisors of n (numbers less than n which divide evenly into n).
// If d(a) = b and d(b) = a, where a  b, then a and b are an amicable pair and each of a and b are called amicable numbers.

// For example, the proper divisors of 220 are 1, 2, 4, 5, 10, 11, 20, 22, 44, 55 and 110; therefore d(220) = 284. The proper divisors of 284 are 1, 2, 4, 71 and 142; so d(284) = 220.
// d(a) == b && d(b) == a
// sum[a] == sum[sum[a]]

// Evaluate the sum of all the amicable numbers under 10000.

class P21 {
    
    public static void main(String[] args) {
	
	int roof = 10000; 
	
	int sum[] = new int[roof]; 
	
	//Compute the sum of proper divisors for every number under 10000. 
	for(int i = 0; i < roof; i++)
	    sum[i] = d(i); 	    	    	
	
	int res = 0; 
	
	for(int a = 0; a < roof; a++){
	    if(sum[a] < roof){
		int b = sum[a]; 
		if(b != a && sum[b] == a) {
		    System.out.println(a+ " and "+ b+" are amicable "); 
		    System.out.println("sum[a] : "+sum[a]); 
		    System.out.println("sum[b] : "+sum[b]); 
		    res += (b + a);
		}
	    }
	}	
	//While looping over the numbers in [0,roof], each pair a,b will be recognized as amicable two times (when i = a, and when i = b)	
	//Divide by two to count them only once.
	System.out.println("Sum amicable numbers: "+(res / 2)); 
    }

    // Sum of proper divisors of n. 
    public static int d(int num){
	
	int res = 0; 
	
	for(int i = 1; i < num; i++)
	    if(num % i == 0) 
		res += i; 	    
	return res; 			
    }
}

File: euler/p21.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-

"""
Project Euler, problem 21: Amicable numbers
(http://projecteuler.net/problem=21).

Let d(n) be defined as the sum of proper divisors of n (numbers less than n
which divide evenly into n).

If d(a) = b and d(b) = a, where a != b, then a and b are an amicable pair and
each of a and b are called amicable numbers.

For example, the proper divisors of 220 are 1, 2, 4, 5, 10, 11, 20, 22, 44, 55
and 110; therefore d(220) = 284. The proper divisors of 284 are 1, 2, 4, 71 and
142; so d(284) = 220.

d(a) == b && d(b) == a
sum[a] == sum[sum[a]]

Evaluate the sum of all the amicable numbers under 10000.
"""

def get_proper_divisors(num):
    """
    The proper divisors are the numbers less than num which divide evenly into
    num
    """
    return [i for i in range(1, num) if num % i == 0]

def main():
    """
    Compute all the amicable numbers, storing them on a dictionary without
    duplicates (every couple of amicable numbres is stored only once.
    """

    amicable = {}
    sum_amicable = 0

    for i in range(2, 10000):
        if i in amicable.values():
            continue

        divisors_sum = sum(get_proper_divisors(i))
        if i != divisors_sum and i == sum(get_proper_divisors(divisors_sum)):
            amicable[i] = divisors_sum
            sum_amicable += (i + divisors_sum)

    print "The sum of the amicable numbers is %d " % sum_amicable

if __name__ == "__main__":
    main()
